nifi-nar-bundles/nifi-windows-event-log-bundle/nifi-windows-event-log-processors/src/main/java/org/apache/nifi/processors/windows/event/log/ConsumeWindowsEventLog.java

```diff
@@ -0,0 +1,282 @@
+            wEvtApiError = e;
+            return null;
+        }
+    }
+
+    private Kernel32 loadKernel32() {
+        try {
+            return Kernel32.INSTANCE;
+        } catch (Throwable e) {
+            kernel32Error = e;
+            return null;
+        }
+    }
+
+    /**
+     * Register subscriber via native call
+     *
+     * @param context the process context
+     */
+    @OnScheduled
+    public void onScheduled(ProcessContext context) throws Exception {
+        String channel = context.getProperty(CHANNEL).getValue();
+        String query = context.getProperty(QUERY).getValue();
+
+        renderedXMLs = new LinkedBlockingQueue<>(context.getProperty(MAX_EVENT_QUEUE_SIZE).asInteger());
+        provenanceUri = new URI("winlog", name, "/" + channel, query, null).toASCIIString();
+
+        evtSubscribeCallback = new EventSubscribeXmlRenderingCallback(getLogger(), s -> {
+            try {
+                renderedXMLs.put(s);
+            } catch (InterruptedException e) {
+                throw new IllegalStateException("Got interrupted while waiting to add to queue.", e);
+            }
+        }, context.getProperty(MAX_BUFFER_SIZE).asInteger(), wEvtApi, kernel32, errorLookup);
+
+        subscriptionHandle = wEvtApi.EvtSubscribe(null, null, channel, query, null, null,
+                evtSubscribeCallback, WEvtApi.EvtSubscribeFlags.SUBSCRIBE_TO_FUTURE);
+        if (subscriptionHandle == null || subscriptionHandle.getPointer() == null) {
+            throw new Exception("Unable to subscribe with provided parameters, received the following error code: "
+                    + errorLookup.getLastError());
+        }
+    }
+
+    /**
+     * Cleanup
+     */
+    @OnStopped
+    public void stop() {
+        wEvtApi.EvtClose(subscriptionHandle);
+        subscriptionHandle = null;
+        evtSubscribeCallback = null;
+        if (!renderedXMLs.isEmpty()) {
+            if (sessionFactory != null) {
+                getLogger().info("Finishing processing leftover events");
+                ProcessSession session = sessionFactory.createSession();
+                processQueue(session);
+            } else {
+                throw new ProcessException("Stopping the processor but there is no ProcessSessionFactory stored and there are messages in the internal queue. Removing the processor now will " +
+                        "clear the queue but will result in DATA LOSS. This is normally due to starting the processor, receiving events and stopping before the onTrigger happens. The messages " +
+                        "in the internal queue cannot finish processing until until the processor is triggered to run.");
+            }
+        }
+        sessionFactory = null;
+        provenanceUri = null;
+        renderedXMLs = null;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSessionFactory sessionFactory) throws ProcessException {
+        this.sessionFactory = sessionFactory;
```

Comment:
They're both correct but checking for null is highly optimized by the compiler and reassigning the value leads to unnecessary GC because of losing reference to the previous assignment.